File: frontend/src/types/prediction.ts
export type RiskTier = 'low' | 'moderate' | 'high' | 'critical'

export interface Contributor {
  feature: string
  rank: number
  shap_value: number
  raw_value: number | string | boolean | null
  sentence: string
}

export interface RiskPrediction {
  probability: number
  risk_tier: RiskTier
  algorithm: string
  top_factors: Contributor[]
}

export interface GpaPrediction {
  predicted_gpa: number
  predicted_cgpa: number
  interval_low: number
  interval_high: number
  algorithm: string
  top_factors: Contributor[]
}

export interface CourseScorePrediction {
  course_id: number
  course_code: string
  predicted_score: number
  algorithm: string
  top_factors: Contributor[]
}

export interface StudentPredictionResponse {
  student_id: number
  session: string
  semester: string
  risk: RiskPrediction | null
  gpa: GpaPrediction | null
  course_scores: CourseScorePrediction[]
}

export interface AtRiskItem {
  student_id: number
  matric_no: string
  full_name: string
  department: string
  level: number
  risk_tier: RiskTier
  probability: number
  adviser_id: number | null
}

export interface AtRiskResponse {
  items: AtRiskItem[]
  total: number
}

export interface BatchPredictionRow {
  student_id: number
  matric_no: string
  risk_tier: RiskTier | null
  probability: number | null
  predicted_gpa: number | null
  error: string | null
}

export interface BatchPredictionResponse {
  total_requested: number
  total_succeeded: number
  results: BatchPredictionRow[]
}

File: backend/app/schemas/prediction.py
"""Prediction request/response schemas."""

from pydantic import BaseModel

from app.models.enums import RiskTier


class Contributor(BaseModel):
    feature: str
    rank: int
    shap_value: float
    raw_value: float | str | bool | None
    sentence: str


class RiskPrediction(BaseModel):
    probability: float
    risk_tier: RiskTier
    algorithm: str
    top_factors: list[Contributor]


class GpaPrediction(BaseModel):
    predicted_gpa: float
    predicted_cgpa: float
    interval_low: float
    interval_high: float
    algorithm: str
    top_factors: list[Contributor]


class CourseScorePrediction(BaseModel):
    course_id: int
    course_code: str
    predicted_score: float
    algorithm: str
    top_factors: list[Contributor]


class StudentPredictionResponse(BaseModel):
    student_id: int
    session: str
    semester: str
    risk: RiskPrediction | None
    gpa: GpaPrediction | None
    course_scores: list[CourseScorePrediction]


class BatchPredictionRequest(BaseModel):
    student_ids: list[int] | None = None


class BatchPredictionRow(BaseModel):
    student_id: int
    matric_no: str
    risk_tier: RiskTier | None
    probability: float | None
    predicted_gpa: float | None
    error: str | None = None


class BatchPredictionResponse(BaseModel):
    total_requested: int
    total_succeeded: int
    results: list[BatchPredictionRow]


class AtRiskItem(BaseModel):
    student_id: int
    matric_no: str
    full_name: str
    department: str
    level: int
    risk_tier: RiskTier
    probability: float
    adviser_id: int | None


class AtRiskResponse(BaseModel):
    items: list[AtRiskItem]
    total: int


class ExplanationResponse(BaseModel):
    student_id: int
    task: str
    algorithm: str
    sentences: list[str]
    contributors: list[Contributor]
